README.md
# Universal Validator

---

[![NPM version](https://badge.fury.io/js/univ.png)](https://www.npmjs.com/package/univ)
[![spm package](http://spmjs.io/badge/univ)](http://spmjs.io/package/univ)
[![Build Status](https://secure.travis-ci.org/webforms/univ.png?branch=master)](https://travis-ci.org/webforms/univ)
[![Coverage Status](https://coveralls.io/repos/webforms/univ/badge.png?branch=master)](https://coveralls.io/r/webforms/univ)

Universal Validator.

---

## Install

via spm:

```
$ spm install univ
```

via npm:

```
$ npm install univ
```

## Usage

```js
var Validator = require('univ');

var rules = {
  "username": {
    type: "email",
    required: true,
    custom: function(value){
      // Async validate.
      return new Promise(function(resolve, reject){
          $.post("/check-username-available", {username:value}, function(result) {
            resolve(result.available === "YES")
          }, "json")
      })
    }
  },
  "password": {
    type: "password",
    required: true,
    minlength: 6,
    maxlength: 30
  },
  "re-password": {
    type: "password",
    required: true,
    minlength: 6,
    maxlength: 30,
    custom: function(value){
      // Sync validate.
      return value === this.data("password");
    }
  }
};

var validator = new Validator(rules);
var result = yield validator.validate({
  "username": "hotoo@email.address",
  "password": "PassWord",
  "checkbox": ["check-0", "check-1"]
});
```

## API

### Validator(ruler)

constructor, new a validator by rulers.

```js
{
  // rule name.
  "name": {
    type: {TypeEnum(
      text,password, search,textarea,
      radio,checkbox,
      select-one,select-multiple,
      number,range, email,url,tel,color,
      date,week,month,time,datetime,datetime-local,
      file,
      submit,button,image,
      hidden
    )},
    required: {Boolean},
    max: {Number},
    min: {Number},
    maxlength: {Number},
    minlength: {Number},
    pattern: {RegExp},
    multiple: {Boolean},
    step: {Number},
    accept: {Array<String>},
    custom: {Function}
  },
  "other-name": {
    // ...
  },
  // ...
}
```


### Univ.rule(name, rule)

Set or get a custom rule.

```js
validator.rule("isBankCard", function(values){
  return true;
});
```


### univ.validate(data)

data:

```js
univ.validate(
  {
    "name": "value",
    "other-name": ["item-1", "item-2"],
    // ...
  }
);
```

## Rule

### {Boolean} required

Set field is required or not.

### {String} type

Set field type, type value is a enum list(same html5 type):

* text
* password
* search
* textarea
* radio
* checkbox
* select-one
* select-multiple
* number
* range
* email
* url
* tel
* color
* date
* week
* month
* time
* datetime
* datetime-local
* file
* submit
* button
* image
* hidden

### {Number} min

* For `[type=number]`, input data must be great than the min value.
* For `[type=date]` and so on, input data

## Events

### valid

单个数据项通过校验，数据合法有效。

```js
validator.on("valid", function(name, value, validity){
  console.log("Field [name=" + name + "] Passed.");
});
```

### invalid

单个数据项未通过校验，数据无效、不合法。

```js
validator.on("valid", function(name, value, validity){
  console.log("Field [name=" + name + "] Failed.");
});
```

### complete

所有数据校验完成。数据是否全部通过校验，则视事件处理函数的对应参数。

```js
// @param {Boolean} certified.
validator.on("complete", function(certified){
  console.log("Form validation", certified ? "Passed" : "Failed");
});
```

### error

校验过程中出现异常，则抛出 `error` 事件。

```js
// @param {Error} error
validator.on("error", function(error){
});
```


HISTORY.md
# CHANGELOG

---

## 0.2.3 (2016-01-08)

* Update mobile number rule for China. fixed #17, #18.

## 0.2.2 (2015-02-09)

* Fixed: #13, `type=email` not support `-` character in domain name. base on v0.2.1

## 0.2.1 (2015-01-29)

* #12 解决 Promise 带来的性能问题。

## 0.2.0 (2015-01-28)

* Update Promise support.
* Delete moment deps.

## 0.1.7 (2015-02-09)

* Fixed: #13, `type=email` not support `-` character in domain name.

## 0.1.6 (2014-12-24)

* Add `validity` feature, #8.

## 0.1.5 (2014-08-01)

* Fixed #7, emit invalid event for required.

## 0.1.4 (2014-06-26)

* Update `[type=tel]` both support tel and mobile number.

## 0.1.3 (2014-06-25)

* Fixed `isNaN(null)===true` for `verifyMin()` and `verifyMax()`.
* Add test case for type=number.

## 0.1.2 (2014-06-25)

* Fixed `minlength` and `maxlength`.
* Fixed `verifyMaxLengthList()`

## 0.1.1 (2014-06-16)

* Fixed #5, `minlength` and `maxlength` rule.
* Fxied #6, add `minlimit` and `maxlimit` rule.

## 0.1.0 (2014-06-10)

* `new` It is the first version of universal validator.


tests/date-util-spec.js
var expect = require('expect.js');
var dateUtil = require('../date-util');

function typeOf(type) {
  return function(object) {
    return Object.prototype.toString.call(object) === '[object ' + type + ']'
  }
}
var isArray = typeOf("Array")
var isNumber = typeOf("Number")

function isAbsNaN(object){
  return isNumber(object) && isNaN(object)
}

function each(list, handler) {
  for(var i=0,l=list.length; i<l; i++){
    handler(list[i], i)
  }
}

describe("dateUtil", function(){

  var testcases_getDateOfWeek = [
   [ [2015, 1, 0], [2014, 11, 29] ],
   [ [2013, 1, 0], [2012, 11, 31] ],
   [ [1999, 1, 0], [1999, 0, 4] ]
  ]

  each(testcases_getDateOfWeek, function(testcase){
    it('getDateOfWeek(' + testcase[0].join(',') + ') == ' +
      testcase[1][0] + '-' + (testcase[1][1]+1) + '-' + testcase[1][2], function(){

      var date = dateUtil.getDateOfWeek.apply(dateUtil, testcase[0])
      expect(date.getFullYear()).to.equal(testcase[1][0])
      expect(date.getMonth()).to.equal(testcase[1][1])
      expect(date.getDate()).to.equal(testcase[1][2])
    })
  })

  var testcases_parseDate = [

    // DATETIME.
    [ "2009-01-01T00:00:00", [2009, 0, 1, 0, 0, 0, 0] ],
    [ "0000-01-01T00:00:00", [0, 0, 1, 0, 0, 0, 0] ],
    [ "-0001-01-01T00:00:00", [-1, 0, 1, 0, 0, 0, 0] ],
    // DATETIME: month boundary test.
    [ "2009-00-01T00:00:00", NaN ],
    [ "2009-12-01T00:00:00", [2009, 11, 1, 0, 0, 0, 0] ],
    [ "2009-13-01T00:00:00", NaN ],
    // DATETIME: date boundary test.
    [ "2009-01-00T00:00:00", NaN ],
    [ "2009-01-31T00:00:00", [2009, 0, 31, 0, 0, 0, 0] ],
    [ "2009-01-32T00:00:00", NaN ],
    // DATETIME: hours boundary test.
    [ "2009-01-01T-1:00:00", NaN ],
    [ "2009-01-01T23:00:00", [2009, 0, 1, 23, 0, 0, 0] ],
    [ "2009-01-01T24:00:00", NaN ],
    // DATETIME: minutes boundary test.
    [ "2009-01-01T00:-1:00", NaN ],
    [ "2009-01-01T00:59:00", [2009, 0, 1, 0, 59, 0, 0] ],
    [ "2009-01-01T00:60:00", NaN ],
    // DATETIME: seconds boundary test.
    [ "2009-01-01T00:00:-1", NaN ],
    [ "2009-01-01T00:00:59", [2009, 0, 1, 0, 0, 59, 0] ],
    [ "2009-01-01T00:00:60", NaN ],

    // DATETIME: space split date and time.
    //![ "2009-01-01 00:00:00", [2009, 0, 1, 0, 0, 0, 0] ],

    // DATETIME: no seconds.
    //![ "2009-01-01 00:00", [2009, 0, 1, 0, 0, 0, 0] ],
    [ "0000-01-01T00:00", [0, 0, 1, 0, 0, 0, 0] ],
    [ "-0001-01-01T00:00", [-1, 0, 1, 0, 0, 0, 0] ],
    // DATETIME: month boundary test.
    [ "2009-00-01T00:00", NaN ],
    [ "2009-12-01T00:00", [2009, 11, 1, 0, 0, 0, 0] ],
    [ "2009-13-01T00:00", NaN ],
    // DATETIME: date boundary test.
    [ "2009-01-00T00:00", NaN ],
    [ "2009-01-31T00:00", [2009, 0, 31, 0, 0, 0, 0] ],
    [ "2009-01-32T00:00", NaN ],
    // DATETIME: hours boundary test.
    [ "2009-01-01T-1:00", NaN ],
    [ "2009-01-01T23:00", [2009, 0, 1, 23, 0, 0, 0] ],
    [ "2009-01-01T24:00", NaN ],
    // DATETIME: minutes boundary test.
    [ "2009-01-01T00:-1", NaN ],
    [ "2009-01-01T00:59", [2009, 0, 1, 0, 59, 0, 0] ],
    [ "2009-01-01T00:60", NaN ],

    // DATE.
    [ "2009-01-00", NaN ],
    [ "2009-01-01", [2009, 0, 1, 0, 0, 0, 0] ],
    [ "2009-01-32", NaN ],
    [ "2009-02-28", [2009, 1, 28, 0, 0, 0, 0] ],
    [ "2009-02-29", NaN ],
    [ "2009-12-31", [2009, 11, 31, 0, 0, 0, 0] ],
    [ "2009-12-32", NaN ],
    [ "0100-12-31", [100, 11, 31, 0, 0, 0, 0] ],
    [ "0099-12-31", [99, 11, 31, 0, 0, 0, 0] ],
    [ "0001-12-31", [1, 11, 31, 0, 0, 0, 0] ],
    [ "0001-01-01", [1, 0, 1, 0, 0, 0, 0] ],
    [ "0000-01-01", [0, 0, 1, 0, 0, 0, 0] ],
    [ "-0000-01-01", [0, 0, 1, 0, 0, 0, 0] ],
    [ "-0001-01-01", [-1, 0, 1, 0, 0, 0, 0] ],

    [ "2009-00-01", NaN ],
    [ "2009-13-01", NaN ],
    [ "2009-100-01", NaN ],
    [ "2009-001-01", NaN ],
    [ "2009-01-001", NaN ],

    // MONTH.
    [ "2009-01", [2009, 0, 1, 0, 0, 0, 0] ],
    [ "2009-12", [2009, 11, 1, 0, 0, 0, 0] ],
    [ "0100-12", [100, 11, 1, 0, 0, 0, 0] ],
    [ "0099-12", [99, 11, 1, 0, 0, 0, 0] ],
    [ "0001-12", [1, 11, 1, 0, 0, 0, 0] ],
    [ "0001-01", [1, 0, 1, 0, 0, 0, 0] ],
    [ "0000-01", [0, 0, 1, 0, 0, 0, 0] ],
    [ "-0000-01", [0, 0, 1, 0, 0, 0, 0] ],
    [ "-0001-01", [-1, 0, 1, 0, 0, 0, 0] ],

    [ "2009-00", NaN ],
    [ "2009-13", NaN ],
    [ "2009-100", NaN ],
    [ "2009-001", NaN ],

    // TIME
    [ "00:00:00", [1900, 0, 1, 0, 0, 0, 0] ],
    [ "00:00:59", [1900, 0, 1, 0, 0, 59, 0] ],
    [ "00:00:60", NaN ],
    [ "00:59:00", [1900, 0, 1, 0, 59, 0, 0] ],
    [ "00:60:00", NaN ],
    [ "23:00:00", [1900, 0, 1, 23, 0, 0, 0] ],
    [ "24:00:00", NaN ],

    [ "00:00", [1900, 0, 1, 0, 0, 0, 0] ],
    [ "00:59", [1900, 0, 1, 0, 59, 0, 0] ],
    [ "00:60", NaN ],
    [ "23:00", [1900, 0, 1, 23, 0, 0, 0] ],
    [ "24:00", NaN ],

    // WEEK
    [ "2009-W01-1", [2008, 11, 29, 0, 0, 0, 0] ],
    [ "2009-W53-7", [2010, 0, 3, 0, 0, 0, 0] ],
    [ "2009-W54-1", NaN ],
    [ "2009-W01-0", NaN ],
    [ "2009-W01-8", NaN ],
    [ "2009-W541", NaN ],
    [ "2009-W010", NaN ],
    [ "2009-W018", NaN ],
    [ "2015-W01", [2014, 11, 29, 0, 0, 0, 0] ],
    [ "2015-W011", [2014, 11, 29, 0, 0, 0, 0] ],
    [ "2015-W01-1", [2014, 11, 29, 0, 0, 0, 0] ],

  ]

  each(testcases_parseDate, function(testcase){
    var desc = 'parseDate(' + testcase[0] + ') == ' +
      (isAbsNaN(testcase[1]) ? "NaN" :
        testcase[1][0] + '-' + (testcase[1][1]+1) + '-' + testcase[1][2] +
        "T" + testcase[1][3] + ":" + testcase[1][4] + ":" + testcase[1][5])

    it(desc, function(){

      var date = dateUtil.parseDate(testcase[0])

      if (isAbsNaN(testcase[1])) {
        expect(isAbsNaN(date)).to.equal(true)
      } else {
        expect("year:"+date.getFullYear()).to.equal("year:"+testcase[1][0])
        expect("month:"+date.getMonth()).to.equal("month:"+testcase[1][1])
        expect("date:"+date.getDate()).to.equal("date:"+testcase[1][2])
        expect("hours:"+date.getHours()).to.equal("hours:"+testcase[1][3])
        expect("minutes:"+date.getMinutes()).to.equal("minutes:"+testcase[1][4])
        expect("seconds:"+date.getSeconds()).to.equal("seconds:"+testcase[1][5])
      }
    })
  })

  var testcases_getWeeksOfYear = [
    [2015, 53],
    [2014, 52],
    [2013, 52],
    [2012, 52],
    [2011, 52],
    [2010, 52],
    [2009, 53],
    [2008, 52],
    [2007, 52],
    [2006, 52],
    [2005, 52],
    [2004, 53],
    [2003, 52],
    [2002, 52],
    [2001, 52],
    [2000, 52],
    [1999, 52],
    [1998, 53],
    [1997, 52],
    [1996, 52],
    [1995, 52],
    [1994, 52],
    [1993, 52],
    [1992, 53],
    [1991, 52],
    [1990, 52],
    [1989, 52],
    [1988, 52],
    [1987, 53],
    [1986, 52],
    [1985, 52],
    [1984, 52],
    [1983, 52],
    [1982, 52],
    [1981, 53],
    [1980, 52],
    [1979, 52],
    [1978, 52],
    [1977, 52],
    [1976, 53]
  ]

  each(testcases_getWeeksOfYear, function(testcase){
    var desc = 'getWeeksOfYear(' + testcase[0] + ') == ' + testcase[1]

    it(desc, function(){

      var weeks = dateUtil.getWeeksOfYear(testcase[0])
      expect(weeks).to.equal(testcase[1])

    })
  })


  var testcases_compareDate = [
    // date : date
    [ "1900-05-05", "1900-05-05", 0 ],
    [ "1901-05-05", "1900-05-05", 1 ],
    [ "1900-06-01", "1900-05-05", 1 ],
    [ "1900-05-06", "1900-05-05", 1 ],
    [ "1899-05-05", "1900-05-05", -1 ],
    [ "1900-04-05", "1900-05-05", -1 ],
    [ "1900-05-04", "1900-05-05", -1 ],

    // datetime : datetime
    [ "1900-05-05T05:05:05", "1900-05-05T05:05:05", 0 ],
    [ "1901-05-05T05:05:05", "1900-05-05T05:05:05", 1 ],
    [ "1900-06-05T05:05:05", "1900-05-05T05:05:05", 1 ],
    [ "1900-05-06T05:05:05", "1900-05-05T05:05:05", 1 ],
    [ "1900-05-05T06:05:05", "1900-05-05T05:05:05", 1 ],
    [ "1900-05-05T05:06:05", "1900-05-05T05:05:05", 1 ],
    [ "1900-05-05T05:05:06", "1900-05-05T05:05:05", 1 ],
    [ "1899-05-05T05:05:05", "1900-05-05T05:05:05", -1 ],
    [ "1900-04-05T05:05:05", "1900-05-05T05:05:05", -1 ],
    [ "1900-05-04T05:05:05", "1900-05-05T05:05:05", -1 ],
    [ "1900-05-05T04:05:05", "1900-05-05T05:05:05", -1 ],
    [ "1900-05-05T05:04:05", "1900-05-05T05:05:05", -1 ],
    [ "1900-05-05T05:05:04", "1900-05-05T05:05:05", -1 ],
    [ "1900-05-05T05:05", "1900-05-05T05:05", 0 ],
    [ "1901-05-05T05:05", "1900-05-05T05:05", 1 ],
    [ "1900-06-05T05:05", "1900-05-05T05:05", 1 ],
    [ "1900-05-06T05:05", "1900-05-05T05:05", 1 ],
    [ "1900-05-05T06:05", "1900-05-05T05:05", 1 ],
    [ "1900-05-05T05:06", "1900-05-05T05:05", 1 ],
    [ "1899-05-05T05:05", "1900-05-05T05:05", -1 ],
    [ "1900-04-05T05:05", "1900-05-05T05:05", -1 ],
    [ "1900-05-04T05:05", "1900-05-05T05:05", -1 ],
    [ "1900-05-05T04:05", "1900-05-05T05:05", -1 ],
    [ "1900-05-05T05:04", "1900-05-05T05:05", -1 ],

    // time : time
    [ "05:05:05", "05:05:05", 0 ],
    [ "06:05:05", "05:05:05", 1 ],
    [ "05:06:05", "05:05:05", 1 ],
    [ "05:05:06", "05:05:05", 1 ],
    [ "04:05:05", "05:05:05", -1 ],
    [ "05:04:05", "05:05:05", -1 ],
    [ "05:05:04", "05:05:05", -1 ],
    [ "05:05", "05:05", 0 ],
    [ "06:05", "05:05", 1 ],
    [ "05:06", "05:05", 1 ],
    [ "04:05", "05:05", -1 ],
    [ "05:04", "05:05", -1 ],

    // week : week
    [ "1900-W055", "1900-W055", 0],
    [ "1901-W055", "1900-W055", 1],
    [ "1900-W065", "1900-W055", 1],
    [ "1900-W056", "1900-W055", 1],
    [ "1899-W055", "1900-W055", -1],
    [ "1900-W045", "1900-W055", -1],
    [ "1900-W054", "1900-W055", -1],

    // date : *
    [ "2014-12-29", "2014-12-29T00:00:00", 0 ],
    [ "2014-12-29", "2014-12-29T00:00", 0 ],
    [ "1900-01-01", "00:00:00", 0 ],
    [ "1900-01-01", "00:00", 0 ],
    [ "2014-12-29", "2015-W011", 0 ],
    [ "2014-12-29", "2015-W01", 0 ]
  ]
  each(testcases_compareDate, function(testcase){
    it('compareDate(' + testcase[0] + ',' + testcase[1] + ') == ' + testcase[2], function(){

      expect(dateUtil.compareDate(testcase[0], testcase[1])).to.equal(testcase[2])

    })
  })

  var testcases_isDate = [
    [ "1900-01-01", true],
    [ "1900-12-01", true],
    [ "1900-12-31", true],
    [ "190a-01-01", false],
    [ "1900-00-01", false],
    [ "1900-13-01", false],
    [ "1900-01-00", false],
    [ "1900-01-32", false]
  ]
  each(testcases_isDate, function(testcase){
    it('isDate(' + testcase[0] + ') == ' + testcase[1], function(){

      expect(dateUtil.isDate(testcase[0])).to.equal(testcase[1])

    })
  })

  var testcases_isMonth = [
    [ "1900-01", true],
    [ "1900-12", true],
    [ "190a-01-01", false],
    [ "1900-00", false],
    [ "1900-13", false]
  ]
  each(testcases_isMonth, function(testcase){
    it('isMonth(' + testcase[0] + ') == ' + testcase[1], function(){

      expect(dateUtil.isMonth(testcase[0])).to.equal(testcase[1])

    })
  })

  var testcases_isDateTime = [
    [ "1900-01-01T00:00:00", true],
    [ "1900-12-01T00:00:00", true],
    [ "1900-12-31T00:00:00", true],
    [ "1900-01-01T23:00:00", true],
    [ "1900-01-01T00:59:00", true],
    [ "1900-01-01T00:00:59", true],
    //![ "1900-01-01 00:00:00", true],
    [ "1900-01-01", false],
    [ "1900-00-01T00:00:00", false],
    [ "1900-13-01T00:00:00", false],
    [ "1900-01-00T00:00:00", false],
    [ "1900-01-32T00:00:00", false],
    [ "1900-01-01T24:00:00", false],
    [ "1900-01-01T00:60:00", false],
    [ "1900-01-01T00:00:60", false],
    [ "1900-01-01T00:00", true],
    [ "1900-12-01T00:00", true],
    [ "1900-12-31T00:00", true],
    [ "1900-01-01T23:00", true],
    [ "1900-01-01T00:59", true],
    //![ "1900-01-01 00:00", true],
    [ "1900-00-01T00:00", false],
    [ "1900-13-01T00:00", false],
    [ "1900-01-00T00:00", false],
    [ "1900-01-32T00:00", false],
    [ "1900-01-01T24:00", false],
    [ "1900-01-01T00:60", false]
  ]
  each(testcases_isDateTime, function(testcase){
    it('isDate(' + testcase[0] + ') == ' + testcase[1], function(){

      expect(dateUtil.isDateTime(testcase[0])).to.equal(testcase[1])

    })
  })

  var testcases_isTime = [
    [ "00:00:00", true],
    [ "00:00:00", true],
    [ "00:00:00", true],
    [ "23:00:00", true],
    [ "00:59:00", true],
    [ "00:00:59", true],
    [ "1900-01-01", false],
    [ "24:00:00", false],
    [ "00:60:00", false],
    [ "00:00:60", false],
    [ "00:00", true],
    [ "23:00", true],
    [ "00:59", true],
    [ "24:00", false],
    [ "00:60", false]
  ]
  each(testcases_isTime, function(testcase){
    it('isDate(' + testcase[0] + ') == ' + testcase[1], function(){

      expect(dateUtil.isTime(testcase[0])).to.equal(testcase[1])

    })
  })

  var testcases_isWeek = [
    [ "2015-W01", true],
    [ "2015-W011", true],
    [ "2015-W017", true],
    [ "2015-W00", false],
    [ "2015-W010", false],
    [ "2015-W018", false]
  ]
  each(testcases_isWeek, function(testcase){
    it('isDate(' + testcase[0] + ') == ' + testcase[1], function(){

      expect(dateUtil.isWeek(testcase[0])).to.equal(testcase[1])

    })
  })

});


date-util.js

//                     year       month    date
var RE_DATE = /^([+-]?\d{4,6})\-(\d\d)\-(\d\d)$/
var RE_MONTH = /^([+-]?\d{4,6})\-(\d\d)$/
var RE_TIME = /^(\d\d):(\d\d)(?::(\d\d))?$/
var RE_WEEK = /^([+-]?\d{4,6})-W(\d\d)(?:-?(\d))?$/
var RE_DATETIME = /^([+-]?\d{4,6})\-(\d\d)\-(\d\d)T(\d\d):(\d\d)(?::(\d\d))?(?:[+-]\d\d:\d\d)?Z?$/;

function toInt(string) {
  return parseInt(string, 10)
}

function fixWeekday(weekday){
  return weekday === 0 ? 7 : weekday
}

function isLeapYear(year) {
  return year % 4 === 0 && year % 100 !== 0 && year % 400 === 0;
}

var MONTH_DATES = [31, 28, 31, 30, 31, 30, 30, 31, 30, 31, 30, 31]
// 计算指定年份所在月的天数
function getDaysOfMonth(year, month){
  return MONTH_DATES[month] + (month === 1 && isLeapYear(year) ? 1 : 0)
}

// 计算指定年份第 week 周的周一 0 时的时间。
function getDateOfWeek(year, week, weekday) {
  var dow = new Date(year, 0, 1).getDay()
  var dates = 1 + (week - 1) * 7
  if (1 <= dow && dow <= 4) {
    dates -= dow - 1
  } else {
    dates += 8 - dow
  }
  return new Date(year, 0, dates + (weekday || 1) - 1)
}

// 计算指定年份拥有的周数
// 第 1 周的周四，第 53 周的周四
function getWeeksOfYear(year) {
  var firstWeek = getDateOfWeek(year, 1, 4)
  var lastWeek = getDateOfWeek(year, 53, 4)
  return 52 +
    (firstWeek.getFullYear() !== year ? 1 : 0) +
    (lastWeek.getFullYear() === year ? 1 : 0)
}

function parseDate(string) {
  var match
  var year = 1900
  var month = 0
  var date = 1
  var hours = 0
  var minutes = 0
  var seconds = 0
  var milliseconds = 0

  if (match = RE_DATE.exec(string)) {

    year = toInt(match[1])
    month = toInt(match[2]) - 1
    date = toInt(match[3])

  } else if (match = RE_MONTH.exec(string)) {

    year = toInt(match[1])
    month = toInt(match[2]) - 1

  } else if (match = RE_TIME.exec(string)) {

    hours = toInt(match[1])
    minutes = toInt(match[2])
    seconds = toInt(match[3]) || 0

  } else if (match = RE_WEEK.exec(string)) {

    var y = toInt(match[1])
    var w = toInt(match[2])
    // Unset `match[3]`, day is NaN. NaN < 0 && NaN <0
    // Do'nt set default day by `|| 1`, it will effect weekdayrange [1,7] limit.
    var day = toInt(match[3])
    var maxWeeks = getWeeksOfYear(y)
    if (1 > w || w > maxWeeks || 1 > day || day > 7) {
      return NaN
    }
    var d = getDateOfWeek(y, w, day)
    year = d.getFullYear()
    month = d.getMonth()
    date = d.getDate()

  } else if (match = RE_DATETIME.exec(string)) {

    year = toInt(match[1])
    month = toInt(match[2]) - 1
    date = toInt(match[3])
    hours = toInt(match[4])
    minutes = toInt(match[5])
    seconds = toInt(match[6]) || 0
    milliseconds = toInt(match[7]) || 0

  } else {

    return NaN

  }

  if (0 > month || month > 11 ||
      1 > date || date > getDaysOfMonth(year, month) ||
      0 > hours || hours > 23 ||
      0 > minutes || minutes > 59 ||
      0 > seconds || seconds > 59 ||
      0 > milliseconds || milliseconds > 999
      ) {

    return NaN
  }

  // Use 1900(or another year) and setFullYear for fix new Date(year, ...) not support [0,99] year.
  var dt = new Date(1900, month, date, hours, minutes, seconds, milliseconds)
  dt.setFullYear(year)

  return dt
}

function distanceDate(stringA, stringB) {
  var dateA = parseDate(stringA)
  var dateB = parseDate(stringB)
  if (isNaN(dateA) || isNaN(dateB)) {throw new Error('Invalid Date'); }
  return dateA.getTime() - dateB.getTime()
}

function compareDate(stringA, stringB) {
  var distance = distanceDate(stringA, stringB)
  if (distance === 0) {return 0}
  return distance > 0 ? 1 : -1
}

function isDate(string) {
  var match = string.match(RE_DATE)
  return RE_DATE.test(string) && !isNaN(parseDate(string))
}

function isDateTime(string){
  return RE_DATETIME.test(string) && !isNaN(parseDate(string))
}

function isMonth(string){
  return RE_MONTH.test(string) && !isNaN(parseDate(string))
}

function isTime(string){
  return RE_TIME.test(string) && !isNaN(parseDate(string))
}

function isWeek(string){
  return RE_WEEK.test(string) && !isNaN(parseDate(string))
}

module.exports = {
  getDateOfWeek: getDateOfWeek,
  parseDate: parseDate,
  getWeeksOfYear: getWeeksOfYear,
  compareDate: compareDate,
  distanceDate: distanceDate,
  isDate: isDate,
  isDateTime: isDateTime,
  isMonth: isMonth,
  isTime: isTime,
  isWeek: isWeek
}
